File: EHR-BACKOFFICE-FRONTEND/src/app/admin/structure/material-resource/room/room.service.ts
import { HttpClient, HttpResponse } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Room } from '../room/room.model';
import { Equipment } from '../equipment/equipment.model';

@Injectable({
  providedIn: 'root'
})
export class RoomService {

  private baseUrl = 'http://localhost:8090/parameterization/room';

  constructor(private http: HttpClient) { }

  getAllRooms(): Observable<Room[]> {
    return this.http.get<Room[]>(this.baseUrl);
  }
  createRoom(room: Room): Observable<Room> {
    return this.http.post<Room>(`${this.baseUrl}`, room);
  }
  updateRoom(roomKy: number, room: Room): Observable<Room> {
    return this.http.put<Room>(`${this.baseUrl}/${roomKy}`, room);
  }
  deleteRoom(roomKy: number): Observable<HttpResponse<any>> {
    return this.http.delete<any>(`${this.baseUrl}/${roomKy}`, { observe: 'response' });
  }
  getRoomByNm(roomNm: string): Observable<Room> {
    const url = `${this.baseUrl}/roomName/${roomNm}`;
    return this.http.get<Room>(url);
  }
  getRoomByKy(roomKy: number): Observable<Room> {
    const url = `${this.baseUrl}/${roomKy}`;
    return this.http.get<Room>(url);
  }
  getRoomNameByKey(roomKy: number): Observable<string> {
    const url = `${this.baseUrl}/${roomKy}/name`;
    return this.http.get(url, { responseType: 'text' });
  }
  getChildElements(roomKy: number): Observable<Equipment[]> {
    return this.http.get<Equipment[]>(`${this.baseUrl}/${roomKy}/child-elements`);
  }
  addChildElement(roomKy: number, equipment: Partial<Equipment>): Observable<Equipment> {
    return this.http.post<Equipment>(`${this.baseUrl}/${roomKy}/child-elements`, equipment);
  }
  assignEquipmentsToRoom(roomId: number, equipments: Equipment[]): Observable<Room> {
    const url = `${this.baseUrl}/${roomId}/assignEquipments`;

    return this.http.post<Room>(url, equipments);
  }

  unassignEquipment(roomId: number, equipment: Equipment): Observable<Room> {
    const url = `${this.baseUrl}/${roomId}/unassignEquipment`;
    return this.http.post<Room>(url, equipment);
  }

}

File: BackendPlanner/src/main/java/com/iss4u/BackendPlanner/services/RoomService.java
package com.iss4u.BackendPlanner.services;

import com.iss4u.BackendPlanner.entities.eq.Equipment;
import com.iss4u.BackendPlanner.entities.Room;

import java.util.List;
import java.util.Optional;

public interface RoomService {
    void create(Room room);

    List<Room> retrieveRooms();

    Optional<Room> getRoomByKy(Long roomKy);

    Room getRoomByNm(String roomLabel);

    String getRoomNameByKey(Long roomKy);

    void update(Long roomKy, Room updatedRoom);

    void delete(Long RoomKy);

    List<Equipment> getChildElements(Room room);

    void addChildElement(Room room, Equipment equipment);

    public Room assignEquipmentToRoom(Long roomId, List<Equipment> equipments);

    Room unassignEquipmentFromRoom(Long roomId, Equipment equipment);
}
